File: getHolding.js
require("dotenv").config();
const { PrivyClient } = require("@privy-io/node");
const { ethers } = require("ethers");
const {
  createAuthenticatedRequest,
  hexToPrivateKey,
} = require("./orderly-auth");

// Use native fetch in Node.js 18+ or fallback to node-fetch
let fetch;
if (typeof globalThis !== "undefined" && globalThis.fetch) {
  fetch = globalThis.fetch;
} else {
  try {
    fetch = require("node-fetch");
  } catch (e) {
    throw new Error(
      "fetch is not available. Please install node-fetch: npm install node-fetch"
    );
  }
}

// Configuration
const ORDERLY_API_URL = "https://api.orderly.org"; // or https://testnet-api.orderly.org for testnet
const BROKER_ID = "woofi_pro";

/**
 * Generate Orderly account ID
 * Based on Orderly documentation: accountId = keccak256(abi.encode(address, keccak256(abi.encodePacked(brokerId))))
 *
 * @param {string} address - User wallet address
 * @param {string} brokerId - Broker ID
 * @returns {string} - Account ID as bytes32 hex string
 */
function getAccountId(address, brokerId) {
  const abiCoder = ethers.AbiCoder.defaultAbiCoder();
  // keccak256(abi.encodePacked(brokerId)) - using solidityPackedKeccak256 for ABI-encoded string
  const brokerIdHash = ethers.solidityPackedKeccak256(["string"], [brokerId]);
  // keccak256(abi.encode(address, brokerIdHash))
  return ethers.keccak256(
    abiCoder.encode(["address", "bytes32"], [address, brokerIdHash])
  );
}

/**
 * Get current holding from Orderly account
 * Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/get-current-holding
 *
 * @param {Object} options - Configuration options
 * @param {string} options.walletId - The Privy wallet ID (required)
 * @param {string} options.walletAddress - The wallet address (optional, will be fetched if not provided)
 * @param {boolean} options.all - If true, return all tokens even if balance is empty (optional, default: false)
 * @returns {Promise<Object>} - Holding result
 */
async function getHolding(options = {}) {
  try {
    // Validate environment variables
    const appId = process.env.PRIVY_APP_ID;
    const appSecret = process.env.PRIVY_APP_SECRET;

    if (!appId || !appSecret) {
      throw new Error(
        "Missing required environment variables. Please ensure the following are set in your .env file:\n" +
          "- PRIVY_APP_ID\n" +
          "- PRIVY_APP_SECRET"
      );
    }

    if (!options.walletId) {
      throw new Error("Wallet ID is required. Use --wallet-id <wallet_id>");
    }

    // Get Orderly credentials
    const orderlyKey = process.env.ORDERLY_KEY;
    if (!orderlyKey) {
      throw new Error(
        "ORDERLY_KEY environment variable is required (should be in format 'ed25519:...')"
      );
    }

    const orderlyPrivateKeyHex = process.env.ORDERLY_PRIVATE_KEY;
    if (!orderlyPrivateKeyHex) {
      throw new Error(
        "ORDERLY_PRIVATE_KEY environment variable is required (ed25519 private key in hex format)"
      );
    }

    // Convert hex string to Uint8Array
    const orderlyPrivateKey = hexToPrivateKey(orderlyPrivateKeyHex);

    // Initialize Privy client
    const privy = new PrivyClient({
      appId: appId,
      appSecret: appSecret,
    });

    // Configuration
    const brokerId = BROKER_ID;
    const includeAll = options.all || false;

    // Get wallet address if not provided
    let walletAddress = options.walletAddress;
    if (!walletAddress) {
      console.log("Fetching wallet details...");
      const wallet = await privy.wallets().get(options.walletId);
      walletAddress = wallet.address || wallet.addresses?.[0];
      if (!walletAddress) {
        throw new Error(
          "Could not determine wallet address from wallet object"
        );
      }
      console.log(`   Wallet Address: ${walletAddress}`);
    }

    // Derive Orderly account ID from wallet address and broker ID
    const accountId = getAccountId(walletAddress, brokerId);

    console.log("\nFetching current holding from Orderly...");
    console.log(`   Wallet ID: ${options.walletId}`);
    console.log(`   Wallet Address: ${walletAddress}`);
    console.log(`   Broker ID: ${brokerId}`);
    console.log(`   Account ID: ${accountId}`);
    console.log(`   Include all tokens: ${includeAll}`);

    // Build query parameters
    const queryParams = includeAll ? "?all=true" : "";
    const path = `/v1/client/holding${queryParams}`;

    // Create authenticated request
    const requestConfig = await createAuthenticatedRequest(
      "GET",
      path,
      null,
      accountId,
      orderlyKey,
      orderlyPrivateKey
    );

    // Make the request
    const response = await fetch(`${ORDERLY_API_URL}${path}`, requestConfig);
    const data = await response.json();

    if (!response.ok) {
      throw new Error(
        `Failed to get holding: ${JSON.stringify(data)}`
      );
    }

    if (!data.success) {
      throw new Error(
        `Orderly API returned error: ${JSON.stringify(data)}`
      );
    }

    console.log("\n✅ Holding retrieved successfully!");

    return {
      success: true,
      data: data.data,
      holdings: data.data?.holding || [],
      timestamp: data.timestamp,
      walletAddress: walletAddress,
      accountId: accountId,
    };
  } catch (error) {
    console.error("\n❌ Failed to get holding");
    if (error.message) {
      console.error("Error:", error.message);
    } else {
      console.error(error);
    }
    throw error;
  }
}

/**
 * Main execution function
 */
async function main() {
  try {
    // Parse command line arguments
    const args = process.argv.slice(2);
    const options = {};

    // Simple argument parsing
    for (let i = 0; i < args.length; i++) {
      if (args[i] === "--wallet-id" && args[i + 1]) {
        options.walletId = args[i + 1];
        i++;
      } else if (args[i] === "--wallet-address" && args[i + 1]) {
        options.walletAddress = args[i + 1];
        i++;
      } else if (args[i] === "--all") {
        options.all = true;
      } else if (args[i] === "--help" || args[i] === "-h") {
        console.log(`
Usage: node getHolding.js [options]

Options:
  --wallet-id <id>        Privy wallet ID to use (required)
  --wallet-address <addr>  Wallet address (optional, will be fetched if not provided)
  --all                   Include all tokens even if balance is empty (optional, default: false)
  --help, -h              Show this help message

Examples:
  # Get current holding for a Privy wallet
  node getHolding.js --wallet-id wal_xxx

  # Get holding with wallet address provided
  node getHolding.js --wallet-id wal_xxx --wallet-address 0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb

  # Get all tokens including those with zero balance
  node getHolding.js --wallet-id wal_xxx --all

Environment Variables:
  PRIVY_APP_ID              Privy App ID (required)
  PRIVY_APP_SECRET          Privy App Secret (required)
  ORDERLY_KEY               Orderly public key (ed25519:...) (required)
  ORDERLY_PRIVATE_KEY       Orderly private key in hex format (required)

Note: Orderly account ID is automatically derived from wallet address and broker ID.
        `);
        process.exit(0);
      }
    }

    const result = await getHolding(options);

    // Display holdings in a formatted way
    console.log("\n📊 Current Holdings:");
    console.log("=".repeat(80));

    if (result.holdings.length === 0) {
      console.log("   No holdings found.");
    } else {
      // Calculate total value (if we had prices, but for now just show balances)
      let totalHolding = 0;

      result.holdings.forEach((holding, index) => {
        const available = holding.holding - holding.frozen;
        totalHolding += holding.holding;

        console.log(`\n${index + 1}. ${holding.token}:`);
        console.log(`   Total Holding: ${holding.holding.toLocaleString()}`);
        console.log(`   Frozen: ${holding.frozen.toLocaleString()}`);
        console.log(`   Available: ${available.toLocaleString()}`);
        console.log(`   Pending Short: ${holding.pending_short.toLocaleString()}`);
        console.log(
          `   Updated: ${new Date(holding.updated_time).toLocaleString()}`
        );
      });

      console.log("\n" + "=".repeat(80));
      console.log(`Total Holdings: ${totalHolding.toLocaleString()} (across all tokens)`);
    }

    console.log("\n📝 Summary:");
    console.log(`   Wallet Address: ${result.walletAddress}`);
    console.log(`   Account ID: ${result.accountId}`);
    console.log(`   Number of Tokens: ${result.holdings.length}`);
    console.log(`   Timestamp: ${new Date(result.timestamp).toLocaleString()}`);

    process.exit(0);
  } catch (error) {
    console.error("\n❌ Failed to get holding:", error.message);
    process.exit(1);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

module.exports = { getHolding, getAccountId };

File: get_holding.py
"""
Get current holding from Orderly account
Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/get-current-holding
"""
import os
import sys
import argparse
from dotenv import load_dotenv
from web3 import Web3
from orderly_auth import create_authenticated_request, hex_to_private_key
import requests
from privy_utils import get_account_id, get_wallet_address
from orderly_constants import ORDERLY_API_URL, BROKER_ID
from orderly_db import get_orderly_keys_or_raise

load_dotenv()


def get_holding(wallet_id: str) -> dict:
    """
    Get current holding from Orderly account
    
    Args:
        wallet_id: The Privy wallet ID (required)
        
    Returns:
        Holding result
    """
    app_id = os.getenv("PRIVY_APP_ID")
    app_secret = os.getenv("PRIVY_APP_SECRET")
    
    if not app_id or not app_secret:
        raise ValueError("Missing PRIVY_APP_ID or PRIVY_APP_SECRET")
    if not wallet_id:
        raise ValueError("Wallet ID is required")
    
    # Get Orderly keys from database
    orderly_key, orderly_private_key_hex = get_orderly_keys_or_raise(wallet_id)
    orderly_private_key = hex_to_private_key(orderly_private_key_hex)
    
    # Get wallet address
    print("Fetching wallet details...")
    wallet_address = get_wallet_address(wallet_id, app_id, app_secret)
    print(f"   Wallet Address: {wallet_address}")
    
    account_id = get_account_id(wallet_address, BROKER_ID)
    
    print("\nFetching current holding from Orderly...")
    print(f"   Wallet ID: {wallet_id}")
    print(f"   Wallet Address: {wallet_address}")
    print(f"   Broker ID: {BROKER_ID}")
    print(f"   Account ID: {account_id}")
    
    # Build query parameters
    path = "/v1/client/holding"
    
    # Create authenticated request
    request_config = create_authenticated_request(
        "GET",
        path,
        None,
        account_id,
        orderly_key,
        orderly_private_key
    )
    
    # Make the request
    response = requests.get(
        f"{ORDERLY_API_URL}{path}",
        headers=request_config["headers"]
    )
    
    data = response.json()
    
    if not response.ok:
        raise Exception(f"Failed to get holding: {data}")
    
    if not data.get("success"):
        raise Exception(f"Orderly API returned error: {data}")
    
    print("\n✅ Holding retrieved successfully!")
    
    return {
        "success": True,
        "data": data.get("data"),
        "holdings": data.get("data", {}).get("holding", []),
        "timestamp": data.get("timestamp"),
        "walletAddress": wallet_address,
        "accountId": account_id,
    }


def get_holding_async(wallet_id: str) -> dict:
    """Alternative version of get_holding (kept for compatibility)"""
    app_id = os.getenv("PRIVY_APP_ID")
    app_secret = os.getenv("PRIVY_APP_SECRET")
    
    if not app_id or not app_secret:
        raise ValueError("Missing PRIVY_APP_ID or PRIVY_APP_SECRET")
    if not wallet_id:
        raise ValueError("Wallet ID is required")
    
    # Get Orderly keys from database
    orderly_key, orderly_private_key_hex = get_orderly_keys_or_raise(wallet_id)
    orderly_private_key = hex_to_private_key(orderly_private_key_hex)
    
    print("Fetching wallet details...")
    wallet_address = get_wallet_address(wallet_id, app_id, app_secret)
    print(f"   Wallet Address: {wallet_address}")
    
    account_id = get_account_id(wallet_address, BROKER_ID)
    
    print("\nFetching current holding from Orderly...")
    print(f"   Wallet ID: {wallet_id}")
    print(f"   Wallet Address: {wallet_address}")
    print(f"   Account ID: {account_id}")
    
    path = "/v1/client/holding"
    
    request_config = create_authenticated_request(
        "GET",
        path,
        None,
        account_id,
        orderly_key,
        orderly_private_key
    )
    
    response = requests.get(
        f"{ORDERLY_API_URL}{path}",
        headers=request_config["headers"]
    )
    
    data = response.json()
    
    if not response.ok:
        raise Exception(f"Failed to get holding: {data}")
    
    if not data.get("success"):
        raise Exception(f"Orderly API returned error: {data}")
    
    print("\n✅ Holding retrieved successfully!")
    
    return {
        "success": True,
        "data": data.get("data"),
        "holdings": data.get("data", {}).get("holding", []),
        "timestamp": data.get("timestamp"),
        "walletAddress": wallet_address,
        "accountId": account_id,
    }


def main():
    parser = argparse.ArgumentParser(description="Get current holding from Orderly account")
    parser.add_argument("--wallet-id", required=True, help="Privy wallet ID to use (required)")
    
    args = parser.parse_args()
    
    try:
        result = get_holding(
            wallet_id=args.wallet_id
        )
        
        # Display holdings
        print("\n📊 Current Holdings:")
        print("=" * 80)
        
        if not result["holdings"]:
            print("   No holdings found.")
        else:
            total_holding = 0
            for index, holding in enumerate(result["holdings"], 1):
                available = holding["holding"] - holding["frozen"]
                total_holding += holding["holding"]
                
                print(f"\n{index}. {holding['token']}:")
                print(f"   Total Holding: {holding['holding']:,}")
                print(f"   Frozen: {holding['frozen']:,}")
                print(f"   Available: {available:,}")
                print(f"   Pending Short: {holding.get('pending_short', 0):,}")
                print(f"   Updated: {holding.get('updated_time', 'N/A')}")
            
            print("\n" + "=" * 80)
            print(f"Total Holdings: {total_holding:,} (across all tokens)")
        
        print("\n📝 Summary:")
        print(f"   Wallet Address: {result['walletAddress']}")
        print(f"   Account ID: {result['accountId']}")
        print(f"   Number of Tokens: {len(result['holdings'])}")
        print(f"   Timestamp: {result.get('timestamp', 'N/A')}")
        
        sys.exit(0)
    except Exception as error:
        print(f"\n❌ Failed to get holding: {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

